File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal (root *TreeNode) []int {
	result := make([]int, 0)

	var preOrder func(*TreeNode)
	preOrder = func(node *TreeNode) {
		if node == nil {
			return
		}

		result = append(result, node.Val)
        preOrder(node.Left)
		preOrder(node.Right)
	}

	preOrder(root)

	return result
}

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> answer = new ArrayList<>();

        dfs(answer, root);
        return answer;
    }

    static public void dfs(List<Integer> answer, TreeNode node) {
        if (node == null) {
            return;
        }

        answer.add(node.val);
        dfs(answer, node.left);
        dfs(answer, node.right);
    }
}